File: booking-kit/src/main/java/com/skedgo/android/bookingkit/model/FormField.java
package com.skedgo.android.bookingkit.model;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;

import com.google.gson.annotations.SerializedName;

import java.io.Serializable;

public abstract class FormField implements Parcelable, Serializable {

  public static final int ADDRESS = 5;
  public static final int BOOKINGFORM = 7;
  public static final int DATETIME = 3;
  public static final int LINK = 6;
  public static final int OPTION = 2;
  public static final int STEPPER = 4;
  public static final int STRING = 1;
  public static final int SWITCH = 8;
  public static final Creator<FormField> CREATOR = new Creator<FormField>() {
    @Override
    public FormField createFromParcel(Parcel in) {
      switch (in.readInt()) {
        case STRING:
          return new StringFormField(in);
        case OPTION:
          return new OptionFormField(in);
        case DATETIME:
          return new DateTimeFormField(in);
        case STEPPER:
          return new StepperFormField(in);
        case LINK:
          return new LinkFormField(in);
        case ADDRESS:
          return new AddressFormField(in);
        case BOOKINGFORM:
          return new BookingForm(in);
        case SWITCH:
          return new SwitchFormField(in);
        default:
          return null;
      }
    }

    @Override
    public FormField[] newArray(int size) {
      return new FormField[size];
    }
  };

  @SerializedName("id")
  private String id;
  @SerializedName("title")
  private String title;
  @SerializedName("subtitle")
  private String subtitle;
  @SerializedName("sidetitle")
  private String sidetitle;
  @SerializedName("required")
  private boolean required;
  @SerializedName("readOnly")
  private boolean readOnly;
  @SerializedName("hidden")
  private boolean hidden;
  @SerializedName("type")
  private String type;

  protected FormField(Parcel in) {
    this.id = in.readString();
    this.title = in.readString();
    this.subtitle = in.readString();
    this.sidetitle = in.readString();
    this.required = in.readInt() == 1;
    this.readOnly = in.readInt() == 1;
    this.hidden = in.readInt() == 1;
    this.type = in.readString();
  }

  protected FormField() {

  }

  @Override
  public int describeContents() {
    return 0;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    dest.writeString(id);
    dest.writeString(title);
    dest.writeString(subtitle);
    dest.writeString(sidetitle);
    dest.writeInt(required ? 1 : 0);
    dest.writeInt(readOnly ? 1 : 0);
    dest.writeInt(hidden ? 1 : 0);
    dest.writeString(type);
  }

  public abstract Object getValue();

  @Nullable
  public String getTitle() {
    return title;
  }

  public void setTitle(String title) {
    this.title = title;
  }

  public String getId() {
    return id;
  }

  public void setId(String id) {
    this.id = id;
  }

  public String getType() {
    return type;
  }

  public void setType(String type) {
    this.type = type;
  }

  @Nullable
  public String getSubtitle() {
    return subtitle;
  }

  public void setSubtitle(String subtitle) {
    this.subtitle = subtitle;
  }

  @Nullable
  public String getSidetitle() {
    return sidetitle;
  }

  public void setSidetitle(String sidetitle) {
    this.sidetitle = sidetitle;
  }

  public boolean isRequired() {
    return required;
  }

  public void setRequired(boolean required) {
    this.required = required;
  }

  public boolean isReadOnly() {
    return readOnly;
  }

  public void setReadOnly(boolean readOnly) {
    this.readOnly = readOnly;
  }

  public boolean isHidden() {
    return hidden;
  }

  public void setHidden(boolean hidden) {
    this.hidden = hidden;
  }
}

File: trip-kit-booking/src/main/java/com/skedgo/tripkit/booking/FormField.kt
package com.skedgo.tripkit.booking

import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import com.google.gson.annotations.SerializedName
import java.io.Serializable

abstract class FormField() : Parcelable, Serializable {

    companion object {
        const val STRING = 1
        const val OPTION = 2
        const val DATETIME = 3
        const val STEPPER = 4
        const val ADDRESS = 5
        const val LINK = 6
        const val BOOKINGFORM = 7
        const val SWITCH = 8
        const val PASSWORD = 9
        const val EXTERNAL = 10
        const val QRCODE = 11

        const val ACCESS_TOKEN = "access_token"
        const val EXPIRES_IN = "expires_in"
        const val REFRESH_TOKEN = "refresh_token"
        const val REDIRECT_URI = "redirectUri"
        const val CLIENT_ID = "clientID"
        const val CLIENT_SECRET = "clientSecret"
        const val AUTH_URL = "authURL"
        const val TOKEN_URL = "tokenURL"
        const val SCOPE = "scope"

        @JvmField
        val CREATOR: Parcelable.Creator<FormField> = object : Parcelable.Creator<FormField> {
            override fun createFromParcel(parcel: Parcel): FormField? {
                return when (parcel.readInt()) {
                    PASSWORD -> PasswordFormField(parcel)
                    STRING -> StringFormField(parcel)
                    OPTION -> OptionFormField(parcel)
                    DATETIME -> DateTimeFormField(parcel)
                    STEPPER -> StepperFormField(parcel)
                    LINK -> LinkFormField(parcel)
                    ADDRESS -> AddressFormField(parcel)
                    BOOKINGFORM -> BookingForm(parcel)
                    SWITCH -> SwitchFormField(parcel)
                    QRCODE -> QrFormField(parcel)
                    else -> null
                }
            }

            override fun newArray(size: Int): Array<FormField?> {
                return arrayOfNulls(size)
            }
        }
    }

    @SerializedName("id")
    var id: String? = null

    @SerializedName("title")
    var title: String? = null

    @SerializedName("subtitle")
    var subtitle: String? = null

    @SerializedName("sidetitle")
    var sidetitle: String? = null

    @SerializedName("required")
    var required: Boolean = false

    @SerializedName("readOnly")
    var readOnly: Boolean = false

    @SerializedName("hidden")
    var hidden: Boolean = false

    @SerializedName("type")
    var type: String? = null

    constructor(parcel: Parcel) : this() {
        id = parcel.readString()
        title = parcel.readString()
        subtitle = parcel.readString()
        sidetitle = parcel.readString()
        required = parcel.readInt() == 1
        readOnly = parcel.readInt() == 1
        hidden = parcel.readInt() == 1
        type = parcel.readString()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(id)
        dest.writeString(title)
        dest.writeString(subtitle)
        dest.writeString(sidetitle)
        dest.writeInt(if (required) 1 else 0)
        dest.writeInt(if (readOnly) 1 else 0)
        dest.writeInt(if (hidden) 1 else 0)
        dest.writeString(type)
    }

    abstract fun getValue(): Any?
}
